src/middlewared/middlewared/plugins/catalogs_linux/update.py

```diff
@@ -1,47 +1,117 @@
+        catalogs_dir = os.path.join('/mnt', k8s_dataset, 'catalogs') if k8s_dataset else f'{TMP_IX_APPS_DIR}/catalogs'
+        return {
+            'catalogs_dir': catalogs_dir,
+            'extra': extra or {},
+        }
+
+    @private
+    async def catalog_extend(self, catalog, context):
+        catalog.update({
+            'location': os.path.join(
+                context['catalogs_dir'], convert_repository_to_path(catalog['repository'], catalog['branch'])
+            ),
+            'id': catalog['label'],
+        })
+        extra = context['extra']
+        if extra.get('item_details'):
+            catalog['trains'] = await self.middleware.call(
+                'catalog.items', catalog['label'], {'cache': extra.get('cache', True)},
+            )
+            catalog['healthy'] = all(
+                app['healthy'] for train in catalog['trains'] for app in catalog['trains'][train].values()
+            )
+        return catalog
+
+    @accepts(
+        Dict(
+            'catalog_create',
+            Bool('force', default=False),
+            Str('label', required=True, empty=False),
+            Str('repository', required=True, empty=False),
+            Str('branch', default='master'),
+            register=True,
+        )
+    )
+    async def do_create(self, data):
+        verrors = ValidationErrors()
+        # We normalize the label
+        data['label'] = data['label'].upper()
+
+        if await self.query([['id', '=', data['label']]]):
+            verrors.add('catalog_create.label', 'A catalog with specified label already exists', errno=errno.EEXIST)
+
+        if await self.query([['repository', '=', data['repository']], ['branch', '=', data['branch']]]):
+            for k in ('repository', 'branch'):
+                verrors.add(
+                    f'catalog_create.{k}', 'A catalog with same repository/branch already exists', errno=errno.EEXIST
                 )
 
-        return filter_list(catalogs, filters, options)
+        verrors.check()
+
+        if not data.pop('force'):
+            # We will validate the catalog now to ensure it's valid wrt contents / format
+            path = os.path.join(
+                TMP_IX_APPS_DIR, 'validate_catalogs', convert_repository_to_path(data['repository'], data['branch'])
+            )
+            try:
+                await self.middleware.call('catalog.update_git_repository', {**data, 'location': path}, True)
+                await self.middleware.call('catalog.validate_catalog_from_path', path)
+            finally:
+                await self.middleware.run_in_thread(shutil.rmtree, path, ignore_errors=True)
+
+        await self.middleware.call('datastore.insert', self._config.datastore, data)
+
+        asyncio.ensure_future(self.middleware.call('catalog.sync', data['label']))
+
+        return await self.get_instance(data['label'])
+
+    @accepts(
+        Str('id'),
+    )
+    def do_delete(self, id):
```

Comment:
So you're allowing the OFFICIAL catalog to be deleted?

Seems we should have another attribute in the db (`builtin`) and do not allow builtin to be changed?